arch/arm/core/cortex_m/cpu_idle.c

```diff
@@ -118,11 +118,21 @@ void arch_cpu_atomic_idle(unsigned int key)
 	 * and never touched again.
 	 */
 
-	/*
-	 * Wait for all memory transactions to complete before entering low
-	 * power state.
-	 */
+#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
+	/* No BASEPRI, call wfe directly. (SEVONPEND is set in z_arm_cpu_idle_init()) */
+#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	/* unlock BASEPRI so wfe gets interrupted by incoming interrupts  */
+	__set_BASEPRI(0);
+	__ISB();
+#else
+#error Unsupported architecture
+#endif
+
 	SLEEP_IF_ALLOWED(__WFE);
 
-	arch_irq_unlock(key);
+#if defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	__set_BASEPRI(key);
+	__ISB();
+#endif
+	__enable_irq();
```

Comment:
On platforms without BASEPRI, this will enable interrupts regardless of the `key` parameter value, and this is not what `k_cpu_atomic_idle()` is supposed to do: https://github.com/zephyrproject-rtos/zephyr/blob/6a42ede43b53ceffddeaa814b4bd99284c25240d/include/zephyr/kernel.h#L5863-L5864
I think you should use something like:
```suggestion
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
	if (key != 0U) {
		return;
	}
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
	__set_BASEPRI(key);
	__ISB();
#endif
	__enable_irq();
```

Reply:
Good point. This is ends up essentially being `arch_irq_unlock(key)` with an extra `__enable_irq();` for ARMV7_M and ARMV8_M_MAINLINE.